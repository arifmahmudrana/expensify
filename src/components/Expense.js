import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const Expense = ({description, amount, createdAt, id}) => (
  <div>
    <h3><Link to={`/edit/${id}`}>{description}</Link></h3>
    <p>{numeral(amount/100).format('$0,0.00')} - {moment(createdAt).format('Do MMM, YYYY')}</p>
  </div>
);

export default Expense;