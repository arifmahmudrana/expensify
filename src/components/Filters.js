import React, {Component} from 'react';
import {connect} from 'react-redux';
import { DateRangePicker } from 'react-dates';
import {setTextFilter, sortByAmount, sortByDate, setDate} from '../actions/filters';


export class Filters extends Component {
  state = {
    calendarFocused: null
  };

  onCalendarFocusChange = (calendarFocused) => this.setState(() => ({ calendarFocused }));

  render() {
    return (
      <div className="wrapper">
        <div className="inp-group">
          <div className="inp-group__item">
            <input
              type="text"
              className="inp-txt"
              value={this.props.text}
              onChange={e => this.props.setTextFilter(e.target.value)}
              placeholder="Search expenses"
            />
          </div>
          <div className="inp-group__item">
            <select
              className="select"
              onChange={e => this.props[`sortBy${e.target.value[0].toUpperCase()}${e.target.value.slice(1)}`]()}
              value={this.props.sortBy}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="inp-group__item">
            <DateRangePicker
              startDate={this.props.startDate}
              endDate={this.props.endDate}
              onDatesChange={this.props.setDate}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onCalendarFocusChange}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
              startDateId="start"
              endDateId="end"
            />
          </div>
        </div>        
      </div>
    );
  }
}

const mapStateToProps = ({filters}, props) => ({
  ...filters
});

const mapDispatchToProps = dispatch => ({
  setTextFilter: value => dispatch(setTextFilter(value)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate()),
  setDate: ({ startDate, endDate }) => dispatch(setDate(startDate, endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(Filters);