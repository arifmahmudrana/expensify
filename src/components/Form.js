import React, { Component } from 'react';
import { SingleDatePicker } from 'react-dates';
import moment from 'moment';

class Form extends Component {
  constructor(props) {
    super(props)
    const {description, note, amount, createdAt, id} = this.getExpensesAttributes(props);
    this.state = {
      id,
      description,
      note,
      amount,
      createdAt: moment(createdAt),
      calendarFocused: false,
      error: ''
    };
  }
  getExpensesAttributes(props) {
    const {description = '', note = '', amount = '', createdAt, id = ''} = props;

    return {description, note, amount, createdAt, id};
  }
  componentDidUpdate(prevProps, prevState) {
    if (prevProps.id && this.props.id && prevProps.id !== this.props.id) {
      const {description, note, amount, createdAt, id} = this.getExpensesAttributes(this.props);
      this.setState(() => ({
        id,
        description,
        note,
        amount,
        createdAt: moment(createdAt)
      }));
    }
  }
  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({
      description
    }));
  };
  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({
      note
    }));
  };
  onAmountChange = e => {
    const amount = e.target.value;
    
    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({
        amount
      }));
    }
  };
  onCreatedAtChange = createdAt => createdAt && this.setState(() => ({ createdAt }));
  onCalendarFocusChange = ({ focused }) => this.setState(() => ({ calendarFocused: focused }));
  onSubmit = (e) => {
    e.preventDefault();
    let error = '';
    if (!this.state.description || !this.state.amount) {
      error = 'Please provide description & amount'
    }

    this.setState(() => ({
      error
    }));

    if (!error) {
      const {description, amount, note, createdAt} = this.state
      this.props.onSubmit({
        description: description.trim(), 
        amount: parseFloat(amount, 10) * 100, 
        note: note.trim(), 
        createdAt: createdAt.valueOf()
      });
    }
  }
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onCreatedAtChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onCalendarFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          >
          </textarea>
          <button type="submit">{this.props.id ? 'Edit' : 'Add'} Expense</button>
        </form>
      </div>
    );
  }
}

export default Form;