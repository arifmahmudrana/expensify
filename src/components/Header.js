import React from 'react';
import {Link} from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = ({ startLogout }) => (
  <header className="header">  
    <div className="wrapper header__container">
      <Link className="header__title" to="/dashboard">
        <h1>ExpensiFy</h1>
      </Link>
      <button className="btn btn--link" onClick={startLogout}>Logout</button>
    </div>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);