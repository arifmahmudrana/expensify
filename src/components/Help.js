import React from 'react';

const Help = () => (
  <p>Help</p>
);

export default Help;