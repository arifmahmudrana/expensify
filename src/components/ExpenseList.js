import React from 'react';
import {connect} from 'react-redux';
import Expense from './Expense';
import selectorsExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div>
    {
      props.expenses.length
      ? props.expenses.map(i => <Expense {...i} key={i.id} />)
      : <p>No expenses</p>
    }
  </div>
);

const mapStateToProps = ({expenses, filters}, props) => ({
  expenses: selectorsExpenses(expenses, filters)
});

export default connect(mapStateToProps)(ExpenseList);