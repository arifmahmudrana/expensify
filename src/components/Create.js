import React from 'react';
import Form from './Form';
import {connect} from 'react-redux';
import {addExpense} from '../actions/expenses';

export const Create = ({addExpense}) => (
  <div>
    <h1>Add Expense</h1>
    <Form onSubmit={addExpense} />
  </div>
);

const mapDispatchToProps = (dispatch, {history}) => ({
  addExpense: expense => {
    dispatch(addExpense(expense));
    history.replace('/');
  }
});

export default connect(() => ({}), mapDispatchToProps)(Create);