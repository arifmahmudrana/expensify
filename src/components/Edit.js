import React from 'react';
import {connect} from 'react-redux';
import {updateExpense, deleteExpense} from '../actions/expenses';
import Form from './Form';

export const Edit = ({updateExpense, deleteExpense, expense}) => (
  <div>
    <h1>Edit Expense</h1>
    <Form onSubmit={updateExpense} {...expense} amount={expense.amount/100} />
    <button onClick={deleteExpense}>Delete</button>
  </div>
);

const mapDispatchToProps = (dispatch, {history, match}) => ({
  updateExpense: expense => {
    dispatch(updateExpense({...expense, id: match.params.id}));
    history.replace('/');
  },
  deleteExpense() {
    dispatch(deleteExpense(match.params.id));
    history.replace('/');
  }
});

const mapStateToProps = (state, ownProps) => ({
  expense: state.expenses.find(i => i.id === ownProps.match.params.id)
});

export default connect(mapStateToProps, mapDispatchToProps)(Edit);