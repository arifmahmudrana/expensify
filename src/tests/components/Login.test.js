import React from 'react';
import { shallow } from 'enzyme';
import {Login} from '../../components/Login';

let wrapper, startLogin;

beforeEach(() => {
  startLogin = jest.fn();
  wrapper = shallow(<Login startLogin={startLogin} />);
});

test('Login should render correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should startLogin be called', () => {
  wrapper.find('button').simulate('click');
  expect(startLogin).toHaveBeenCalled();
});