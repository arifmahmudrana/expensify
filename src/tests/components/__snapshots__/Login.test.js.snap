// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render correctly 1`] = `
<div>
  <button
    onClick={[MockFunction]}
  >
    Login
  </button>
</div>
`;
